bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs

```diff
@@ -353,14 +336,17 @@ where
 		}
 
 		// Add the XCM sent in the message to the end of the xcm instruction
-		instructions.extend(message.remote_xcm.0);
+		instructions.extend(message.remote_xcm.clone().0);
 		instructions.push(RefundSurplus);
 		// Refund excess fees to the claimer
 		instructions.push(DepositAsset {
-			assets: Wild(AllOf { id: message.execution_fee.id, fun: WildFungible }),
+			assets: Wild(AllOf { id: message.execution_fee.clone().id, fun: WildFungible }),
 			beneficiary: claimer,
 		});
-		instructions.push(SetTopic(message.topic));
+		// Only add SetTopic if the user didn't already add one
+		if !matches!(&message.remote_xcm.last(), Some(SetTopic(_))) {
+			instructions.push(SetTopic(unique(&message)));
```

Comment:
To reduce suprises, lets follow the convention set by `WithUniqueTopic`, where the XCM message itself is hashed, not the Ethereum message.

I think you can add a helper method `with_unique_topic(xcm)` that just copies these lines: https://github.com/vgeddes/polkadot-sdk/blob/b65c0a3d0382ee423eebf8104d84cec4277b6a07/polkadot/xcm/xcm-builder/src/routing.rs#L47-L53

Reply:
Using that exact code doesn't really work here because we need to extract the topic ID in the `prepare` method, so than we can append it after all the other instructions in `convert`. I've done some refactoring, let me know if this is OK: https://github.com/vgeddes/polkadot-sdk/pull/32/commits/d64dc321b56533a29b38d411a6b9275f3d9155d4

Reply:
I prefer Vincent's approach. I'd assume we can modify the `remote_xcm` in the `prepare` phase(i.e., to include the `SetTopic` instruction if nessessary). Then, there is no need to add it again in the `convert` phase?

Reply:
We can modify the `remote_xcm` in the `prepare` phase, but more [instructions are appended](https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L356) in the `convert` phase. Which means we would have to move the SetTopic instruction twice if we do it in the `prepare` and the `convert`.

Alternatively, we can move the extra instructions added in `convert` to the `prepare` method:
https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L357-L362

Reply:
I think Ron is right here. We shouldn't add an extra  `DepositAsset(beneficiary: claimer)` instruction. The UIs should handle this when building the XCM, and practically all messages will go through the UI.

I think this PR may no longer be necessary with the above change. 

The `SetTopic` at `message.remote_xcm.last()` will be the last instruction in the message. If the topic isn't there, then when BH calls `XcmRouter::send`, a topic will be added automatically with `WithUniqueTopic`.

Reply:
Ok, so I should remove the `DepositAsset(beneficiary: claimer)`? I wonder if its not going to do add more effort for the user, if the UI does not provide this statement. Then the user needs to go through the effort of claiming the trapped assets.

> The SetTopic at message.remote_xcm.last() will be the last instruction in the message. If the topic isn't there, then when BH calls XcmRouter::send, a topic will be added automatically with WithUniqueTopic.

Don't we still need the [`RefundSurplus` instruction here](https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L357)?

Reply:
yes, lets remove all of it. `RefundSurplus` included.

I think 99.9% of users will use a UX (ie. us and turtle), so not too worried.